pkg/ovsdb/ovsdb.go

```diff
@@ -0,0 +1,311 @@
+// Modifications copyright (C) 2017 Che Wei, Lin
+// Copyright 2014 Cisco Systems Inc. All rights reserved.
+// Copyright 2019 Red Hat Inc. All rights reserved.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+// http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ovsdb
+
+import (
+	"errors"
+	"fmt"
+	"github.com/socketplane/libovsdb"
+	"sync"
+)
+
+// OVS driver state
+type OvsBridgeDriver struct {
+	// OVS client
+	ovsClient *libovsdb.OvsdbClient
+
+	// Name of the OVS bridge
+	OvsBridgeName string
+
+	// read/write lock for accessing the cache
+	lock sync.RWMutex
+}
+
+// Create a new OVS driver with Unix socket
+func NewOvsDriverForBridge(bridgeName string) (*OvsBridgeDriver, error) {
+	ovsDriver := new(OvsBridgeDriver)
+
+	ovsDB, err := libovsdb.ConnectWithUnixSocket("/var/run/openvswitch/db.sock")
+	if err != nil {
+		return nil, fmt.Errorf("failed to connect to ovsdb error: %v", err)
+	}
+
+	// Setup state
+	ovsDriver.ovsClient = ovsDB
+	ovsDriver.OvsBridgeName = bridgeName
+
+	bridgeExist, err := ovsDriver.IsBridgePresent(bridgeName)
+	if err != nil {
+		return nil, err
+	}
+
+	if !bridgeExist {
+		return nil, fmt.Errorf("failed to find bridge %s", bridgeName)
+	}
+
+	// Return the new OVS driver
+	return ovsDriver, nil
+}
+
+// Wrapper for ovsDB transaction
+func (self *OvsBridgeDriver) ovsdbTransact(ops []libovsdb.Operation) error {
+	// Perform OVSDB transaction
+	reply, _ := self.ovsClient.Transact("Open_vSwitch", ops...)
+
+	if len(reply) < len(ops) {
+		return errors.New("OVS transaction failed. Unexpected number of replies")
+	}
+
+	// Parse reply and look for errors
+	for i, o := range reply {
+		if o.Error != "" && i < len(ops) {
+			return errors.New("OVS Transaction failed err " + o.Error + "Details: " + o.Details)
+		} else if o.Error != "" {
+			return errors.New("OVS Transaction failed err " + o.Error + "Details: " + o.Details)
+		}
+	}
+
+	// Return success
+	return nil
+}
+
+// **************** OVS driver API ********************
+// Create an internal port in OVS
+func (self *OvsBridgeDriver) CreatePort(infoMap map[string]string, vlanTag uint) error {
+	portUuidStr := infoMap["intfName"]
+	intfUuidStr := fmt.Sprintf("Intf%s", infoMap["intfName"])
+	portUuid := []libovsdb.UUID{{GoUUID: portUuidStr}}
```

Comment:
Oh, I thought we pass these values to OVS as plain strings. Now I see we have to convert them to specific format.

Sorry for the confusion. Would you mind then creating infoMap structure with specific types, or just revert the original code with all parameters listed...